contrib/nginx.md

```diff
@@ -0,0 +1,106 @@
+One benefit of this approach would be that soju doesn't need to be hooked into certbot, nor does it need
+to be reloaded when certificates are updated. Potentially the nginx TLS implementation might be more robust
+but that hasn't been proven :D.
+
+In this example, nginx will also handle serving static files for gamja, from the `/srv/gamja` directory.
+
+**Note:** different Linux distributions use a different configuration structure for nginx, so here we only
+provide the configuration snippets for the [`http`/`server`](https://nginx.org/en/docs/http/ngx_http_core_module.html#server)
+module, the [http proxy](https://nginx.org/en/docs/http/ngx_http_proxy_module.html) configuration and
+the [`stream`/`server`](https://nginx.org/en/docs/stream/ngx_stream_core_module.html) module.
+
+Please refer to your OS documentation to learn how to configure nginx, and where its config files are located.
+
+## http server: websocket proxying
+
+First, configure Soju to listen for websocket on the `/run/soju/ws.sock` unix domain socket:
+```
+…
+listen ws+unix:///run/soju/ws.sock
+…
+```
+
+Then configure nginx to proxy all http traffic for the `/socket` prefix to that unix domain socket,
+by using the `proxy_pass` directives:
+
+```
+# /etc/nginx/sites-enabled/60-soju.conf
+server {
+  server_name chat.example.com;
+  listen 80;
+  listen [::]:80;
+
+  location /socket {
+    proxy_pass http://unix:/run/soju/ws.sock:/socket;
+    proxy_read_timeout 600s;
+    proxy_http_version 1.1;
+    proxy_set_header Upgrade $http_upgrade;
+    proxy_set_header Connection "Upgrade";
+    proxy_set_header X-Forwarded-For $remote_addr;
+    proxy_set_header X-Forwarded-Proto $scheme;
+    proxy_set_header Host $host;
+  }
+
+  root /srv/gamja/;
+}
+```
+All the rest of http requests will be served from staic files in the `/src/gamja` directory
+(you can just copy the gamja `dist/` production files there).
+
+Once you have the above configuration, you can run `sudo certbot --nginx -d chat.example.com`
+and you'll get the configuration file updated for TLS, redirects, and some security headers.
+
+
+## stream server: irc protocol proxying
+
+Nginx, since version 1.9.0, supports the [stream](https://nginx.org/en/docs/stream/ngx_stream_core_module.html)
+module, which can proxy non-http protocols. We can use that to make nginx accepts/handle the tls protocol
+for the irc tcp connections.
+
+Configure Soju to listen on the unix socket `/run/soju/irc.sock` for irc connections:
+```
+…
+listen unix:///run/soju/irc.sock
+…
+```
+
+Now configure nginx to listen on the 1667 tcp port (an an example), and proxy to soju using the "PROXY protocol".
+The `ssl_certificate`, `ssl_certificate_key` and `ssl_dhparam` are copy/pasted from the http server config
+(as edited by certbot).
+```
+# /etc/nginx/streams-enabled/60-soju.conf
+server {
+  listen      1667 ssl;
+  listen      [::]:1667 ssl;
+  proxy_pass unix:/run/soju/irc.sock;
+  proxy_protocol on;
```

Comment:
Note: soju currently ignores the PROXY protocol for Unix sockets. We have `accept-proxy-ips` in the config file but no equivalent for Unix sockets. We need an explicit opt-in on the soju side: if the reverse proxy doesn't enable the PROXY protocol and just passes through bytes, any external user can send a fake PROXY header.

Nothing to change here because it's best to enable this on the nginx side even if the soju side doesn't deal with it yet.